homeassistant/components/compensation/__init__.py

```diff
@@ -0,0 +1,132 @@
+        raise vol.Invalid(
+            f"{CONF_DATAPOINTS} must have at least {value[CONF_DEGREE]+1} {CONF_DATAPOINTS}"
+        )
+
+    return value
+
+
+COMPENSATION_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_ENTITY_ID): cv.entity_id,
+        vol.Required(CONF_DATAPOINTS): vol.All(
+            cv.ensure_list(cv.matches_regex(MATCH_DATAPOINT)),
+        ),
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_ATTRIBUTE): cv.string,
+        vol.Optional(CONF_PRECISION, default=DEFAULT_PRECISION): cv.positive_int,
+        vol.Optional(CONF_DEGREE, default=DEFAULT_DEGREE): vol.All(
+            vol.Coerce(int),
+            vol.Range(min=1, max=7),
+        ),
+        vol.Optional(CONF_UNIT_OF_MEASUREMENT): cv.string,
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {cv.slug: vol.All(COMPENSATION_SCHEMA, datapoints_greater_than_degree)}
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+async def async_setup(hass, config):
+    """Set up the Compensation sensor."""
+    hass.data[DATA_COMPENSATION] = {}
+
+    for compensation, conf in config.get(DOMAIN).items():
+        _LOGGER.debug("Setup %s.%s", DOMAIN, compensation)
+
+        degree = conf[CONF_DEGREE]
+
+        datapoints = []
+        for datapoint in conf[CONF_DATAPOINTS]:
+            match = re.match(MATCH_DATAPOINT, datapoint)
+            # we should always have x and y if the regex validation passed.
+            x_value, y_value = [float(v) for v in match.groups()]
+            datapoints.append((x_value, y_value))
+
+        # get x values and y values from the x,y point pairs
+        x_values, y_values = zip(*datapoints)
+
+        # try to get valid coefficients for a polynomial
+        coefficients = None
+        with np.errstate(all="raise"):
+            with warnings.catch_warnings(record=True) as all_warnings:
+                warnings.simplefilter("always")
+                # try to catch 3 possible errors
+                try:
+                    coefficients = np.polyfit(x_values, y_values, degree)
+                except FloatingPointError as error:
+                    _LOGGER.error(
+                        "Setup of %s encountered an error, %s.",
+                        compensation,
+                        error,
+                    )
+                # raise any warnings
+                for warning in all_warnings:
+                    _LOGGER.warning(
+                        "Setup of %s encountered a warning, %s.",
+                        compensation,
+                        str(warning.message).lower(),
+                    )
+
+        if coefficients is not None:
+            data = {
+                k: v for k, v in conf.items() if k not in [CONF_DEGREE, CONF_DATAPOINTS]
+            }
+            data[CONF_POLYNOMIAL] = np.poly1d(coefficients)
+
+            hass.data[DATA_COMPENSATION][compensation] = data
```

Comment:
doesn't look like you need CONF_POLYNOMIAL here?

Reply:
Probably not, I can change it to just a string 'polynominal', but IIRC I made it into a param in case it needed to be changed/used in multiple spots.